kimchi/src/circuits/lookup/lookups.rs

```diff
@@ -263,13 +261,30 @@ impl<F: Copy> JointLookup<SingleLookup<F>, LookupTableID> {
     }
 }
 
+#[derive(Clone, Serialize, Deserialize)]
 pub enum LookupPattern {
     ChaCha,
     ChaChaFinal,
     LookupGate,
 }
 
 impl LookupPattern {
+    pub fn max_lookups_per_row(&self) -> usize {
+        match self {
+            LookupPattern::ChaCha => 4,
+            LookupPattern::ChaChaFinal => 4,
+            LookupPattern::LookupGate => 3,
+        }
+    }
+
+    pub fn max_joint_size(&self) -> u32 {
+        match self {
+            LookupPattern::ChaCha => 3,
+            LookupPattern::ChaChaFinal => 3,
+            LookupPattern::LookupGate => 2,
+        }
+    }
```

Comment:
both of these functions have hardcoded values, which makes me uneasy

Reply:
I've added a test that checks these constants against the computed values.